src/backend/distributed/commands/extension.c

```diff
@@ -510,25 +510,98 @@ MarkExistingObjectDependenciesDistributedIfSupported()
 	Oid citusTableId = InvalidOid;
 	foreach_oid(citusTableId, citusTableIdList)
 	{
+		if (IsTableOwnedByExtension(citusTableId))
+		{
+			/*
+			 * We let extensions handle its objects by themselves. This includes
+			 * the dependencies that such table has.
+			 */
+			continue;
+		}
+
 		ObjectAddress tableAddress = { 0 };
 		ObjectAddressSet(tableAddress, RelationRelationId, citusTableId);
 
 		/* refrain reading the metadata cache for all tables */
 		if (ShouldSyncTableMetadataViaCatalog(citusTableId))
 		{
-			/* we need to pass pointer allocated in the heap */
-			ObjectAddress *addressPointer = palloc0(sizeof(ObjectAddress));
-			*addressPointer = tableAddress;
+			/*
+			 * All the distributable dependencies of a table should be marked as
+			 * distributed.
+			 */
+			List *distributableDependencyObjectAddresses =
+				GetDistributableDependenciesForObject(&tableAddress);
 
-			/* as of Citus 11, tables that should be synced are also considered object */
-			resultingObjectAddresses = lappend(resultingObjectAddresses, addressPointer);
-		}
+			resultingObjectAddresses =
+				list_concat(resultingObjectAddresses,
+							distributableDependencyObjectAddresses);
 
-		List *distributableDependencyObjectAddresses =
-			GetDistributableDependenciesForObject(&tableAddress);
+			/*
+			 * We mark tables distributed immediately because we also need to mark
+			 * views as distributed. We check whether the views that depend on
+			 * the table has any auto-distirbutable dependencies below. Citus
+			 * currently cannot "auto" distribute tables as dependencies, so we
+			 * mark them distributed immediately.
+			 *
+			 * If a view depends on multiple tables, that view will be marked
+			 * as distributed while it is processed for the last citus
+			 * table.
+			 */
+			MarkObjectDistributedLocally(&tableAddress);
 
-		resultingObjectAddresses = list_concat(resultingObjectAddresses,
-											   distributableDependencyObjectAddresses);
+			/*
```

Comment:
We also support distributing views like `CREATE VIEW basic_view AS SELECT 1;`. In addition to obtaining views depending on any Citus table, should we go over all the views and distribute ones don't depend on any table with an additional loop? Or should we completely skip them to keep code simple?

Reply:
Yes, we should probably support. I pushed a refactor now, and working on adding some tests / discovering whether any further edge cases exists with this approach.

Do you see any?